### PR TITLE
Aftershock Exoskeletons

### PR DESCRIPTION
**Pull request by Drew4484**
Adds modular power armor exoskeletons. A variety of armor plating types as well as  offensive, defensive, and utility modules are available.

#### Summary
Content "Update the modern day exoskeletons for Aftershock"

#### Purpose of change
Update the modern day exoskeletons used in #52322 for Aftershock. Changes will include:

- Battery consumption in line with other Aftershock armor
- Optional utility, offensive, and defensive modules
- Optional work suits be worn in various climates 

#### Describe the solution
Repurpose existing utility exoskeletons, add new capabilities.

#### Describe alternatives you've considered
None

#### Testing

Spawn exoskeleton, functions via UPS power, accepts power supply backpack units. First pass of armor plating (plastic, steel, vacuum carbide) has been tested, all types except steel break as ablative armor and must be replaced or reforged. Steel can be repaired. Modules activate and function properly.

### CODE REVIEW

**Comment by Zireael07** [comment]
Please make the PR title more descriptive... just "Draft" isn't helpful. (Edit button in top right of page)

**Comment by Fris0uman** [comment]
Your summary line should be `Content "Update the modern day exoskeletons for Aftershock"`

File: data/mods/Aftershock/items/tools.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/tools.json:484]: Should split this into a welding tool and a jackhammer. 
That way the jackhammer module could be a decent melee weapon weapon too.

File: data/mods/Aftershock/items/armor/exoskeletons.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/exoskeletons.json:20]: I do something like `"flag_restriction": "gadget_large"` instead

File: data/mods/Aftershock/items/tools.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/tools.json:481]: I dont particularily like this, would be way better to have this be a large  "tool carrier" pocket that lets you fit in a bunch of tools, or alternatively a bunch of non tools at the cost of very minor encumbrance.

**Comment by John-Candlebury** [comment]
So just without actually playtesting this or delving deep into the json definitions:

-There's a .lnk and .font file that you probably committed by mistake, should remove those.
-You also need to add flag definitions to the `flag.h` file, much as you did with `flag.cpp` to get the DDA to compile succesfully.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json:26]:
```suggestion
    "id": "modular_exoskeleton",
```
The old power armor code doesnt like the _off suffix. It would probably be better to fix the old code, but as you mentioned not feelingvery comfortable with C++ I guess you could do this instead.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json:140]:
```suggestion
    "copy-from": "modular_exoskeleton",
```

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json:141]:
```suggestion
    "repairs_like": "modular_exoskeleton",
```

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json:149]:
```suggestion
    "revert_to": "modular_exoskeleton",
```

**Comment by John-Candlebury** [comment]
Theres something strange going on with encumbrance when any of the frame items are equiped with plates.
After succesively powering and unpowering the suit for a few times without taking it off, the suit gains a lot of encumbrance when activated, and has none when inactive.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton.json:229]: Should be covering the hands instead

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json:3]: Most of the items in this file could use copy-from to save lines and also to make it easier to maintain these in the future.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json:86]:
```suggestion
    "name": { "str": "exoskeleton small auxiliary battery" },
```

File: data/mods/Aftershock/items/armor/modular_exos/exo_tool.json
Comment (Mom-Bun) [inline data/mods/Aftershock/items/armor/modular_exos/exo_tool.json:7]: This description feels incomplete to me, mind explaining what you were going for?

I feel like this doesn't need to be re-worked that much description wise, generic is good for now, but I'm curious what is the idea for this description?

File: data/mods/Aftershock/items/armor/modular_exos/exo_gear.json
Comment (Mom-Bun) [inline data/mods/Aftershock/items/armor/modular_exos/exo_gear.json:48]: This might need to be made clear what the exact purpose for this is, maybe somewhat like a construction related? As right now it feels weird but...

 I'd say you could obsolete the old hydraulic gauntlet in aftershock and keep this one but I'd ask the other aftershock maintainers to pitch into this but personally I feel like the old hydraulic gauntlet is a bit bad balance-wise due to the fact its very powerful with no real cost, so having this replace it would be good in my opinion

File: data/mods/Aftershock/items/armor/modular_exos/exo_book.json
Comment (Mom-Bun) [inline data/mods/Aftershock/items/armor/modular_exos/exo_book.json:5]: just noticed you duplicated this, might want to remove one or the other?

**Review by Mom-Bun — COMMENTED** [review]
Sorry for taking a bit to get to this! but this is what I noticed in it, put down my thoughts, and a description for the book possibly if that is alright!

Most of the items in this are very generic so the descriptions didn't need that much for lore-wise I don't think, but I felt like I could fit something in with the manual!

File: data/mods/Aftershock/items/books.json
Comment (Drew4484) [reply data/mods/Aftershock/items/books.json:277]: I like this, I was going to theme it on existing technical manuals but it fits better with the existing robotic brand names.

File: data/mods/Aftershock/items/armor/modular_exos/exo_tool.json
Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/exo_tool.json:7]: Primarily I wanted to have a tool that could work titanium since the standard welders in game can't. Titanium welding is a whole different level of skill already, so the frames would be gated behind a repair proficiency and at least one specialized tool. This could be just a better welder, but until multifuel is an option welding is still abstracted. The vehicle version of this makes more sense as it could provide a shield-gas filled environment and provide solvents/tools to clean the welding sites.

File: data/mods/Aftershock/items/armor/modular_exos/exo_gear.json
Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/exo_gear.json:48]: Replacing the original was the intent with this, though I wasn't sure about making it draw additional power. There aren't a lot of non-combat applications, though it could be useful for demolition. Might fit better as a militarized option.

**Comment by John-Candlebury** [comment]
Im temporarily back from holidays.
 Whats the status for this one? Still want to add stuff or is it ready for final testing?

**Comment by Drew4484** [comment]
> Im temporarily back from holidays.
>  Whats the status for this one? Still want to add stuff or is it ready for final testing?

It should be good to go now. Nothing spawns yet so if there's more changes to be made that can be done in part 2.

**Comment by bombasticSlacks** [comment]
I figure I should say since this has been up, that ablative plates without a transforms rule (something they shatter/deform into) now behave and take damage like normal armor #53713.

That seems like it would actually be a benefit for these plates but just wanted to say.

**Comment by Drew4484** [comment]
> I figure I should say since this has been up, that ablative plates without a transforms rule (something they shatter/deform into) now behave and take damage like normal armor #53713.
> 
> That seems like it would actually be a benefit for these plates but just wanted to say.

That's actually ideal, the plastic and carbide plates should shatter and require replacement, the steel should take damage and be repaired as normal armor. Thanks!

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json:285]: Doesnt seem to be working.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json:371]: Doesnt seem to be working.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json:387]: Renove this one, we should just make an enviromentally sealed frame later on.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_plating.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_plating.json:6]:
```suggestion
    "name": { "str": "steel helmet plating (exosuit)" },
```
Use that naming format for all the other plating items, makes them easier to identify. I prefer the use of the word `exosuit` because its more unique in its uses  than `exoskeleton` (E.X, In game you can potentially mutate an insect exoskeleton, while the word exosuit can only refer to a suit of power armor)

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_plating.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_plating.json:303]:
```suggestion
    "description": "Designed to attach to an exoskeleton's glove, this is a sturdy plate of vacuum-cast carbide.",
```
Fix this error in the other descriptions

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_plating.json
Comment (John-Candlebury) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_plating.json:373]:
```suggestion
    "name": { "str": "vacuum-cast carbide exoskeleton leg plating" },
```
This is improper grammar you should fix all names to be `bodypart in singular` + `plating`

**Review by John-Candlebury — CHANGES_REQUESTED** [review]

**Comment by John-Candlebury** [comment]
Alright some other notes that didnt fit as code change suggestions but still need fixing:

- The encumbrance bug I mentioned above is still there.
- The separate boot and glove items feel supeflous after some play, IMO just remove them and make the leg/arm armor also cover feet and hands.
- The UX for this is really really rough, to the point where I dont think most players will be able to use these without consulting external guides. The worst offenders and ones that really need fixing are:
   - There needs to be a menu or some other dedicated way for you to mount the plates and gadgets into the suit, the current way is too unwieldy. If you want to replace a plating piece with another, you have to drop the one you are wearing and only then can you equip the new one, and also makes it very hard for you to carry extra gadgets without necesarily mounting them itno the suit . It also constantly produces misleading messages about pieces being too heavy to carrry/wield.
   - There should be An `Exosuit  Component` Item category within the inventory,  currently its rather confusing to look at your inventory while wearing a full loadout + exosuit otherwise.

**Comment by Drew4484** [comment]
> Alright some other notes that didnt fit as code change suggestions but still need fixing:
> 
> - The encumbrance bug I mentioned above is still there.
> - The separate boot and glove items feel supeflous after some play, IMO just remove them and make the leg/arm armor also cover feet and hands.
> - The UX for this is really really rough, to the point where I dont think most players will be able to use these without consulting external guides. The worst offenders and ones that really need fixing are:
>    - There needs to be a menu or some other dedicated way for you to mount the plates and gadgets into the suit, the current way is too unwieldy. If you want to replace a plating piece with another, you have to drop the one you are wearing and only then can you equip the new one, and also makes it very hard for you to carry extra gadgets without necesarily mounting them itno the suit . It also constantly produces misleading messages about pieces being too heavy to carrry/wield.
>    - There should be An `Exosuit  Component` Item category within the inventory,  currently its rather confusing to look at your inventory while wearing a full loadout + exosuit otherwise.
> 
> 

I agree that it's not clear what goes where and how to equip items. Making them into a distinct category would help, I can do that. Hands and feet can be parts of arms and legs easy, just loses out on the ability to equip non-power armor boots and gloves which isn't a big loss anyway.

Might make sense to just drop the PSU plating and armor the PSU directly. Loses some customization but simplifies armoring.

I have no idea how possible it would be but hijacking the current bandaging UI would be neat, repairing with patch kits and replacement plates instead of hemostatic powder and bandages.

File: data/mods/Aftershock/items/vehicle_items.json
Comment (Drew4484) [reply data/mods/Aftershock/items/vehicle_items.json:24]: "A vehicle-mounted gantry for supporting an exoskeleton upright during maintenance and repair. Includes the specialized tools needed for complex repairs and will also charge the suit's power when activated from the vehicle's controls."

File: data/mods/Aftershock/items/armor/modular_exos/exo_recipe.json
Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/exo_recipe.json:1]: This is a bit of a placeholder, written before I had the titanium/exo toolkit sorted out. I do want to use the damaged plates as part of creating new ones, maybe another furniture piece/vehicle part.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json
Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json:16]: 20% for the battery, 40% for the ICE. This was an oversight, after talking about the ablative plates the item does need coverage for the plate to apply. So for the battery, 20% coverage for the torso, if it's hit the 100% coverage plate will take that hit. I'll verify with some testing but that's worked for the other plates so far.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json
Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json:371]: Targeted the wrong item, fixed now.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json
Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json:285]: This targeted the wrong item, corrected now.

Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_modules.json:285]: The only active effect is the "PARTIAL_DEAF", like electronic hearing protection. I'd like to duplicate the hearing CBM's effects to complete this item.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json
Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json:3]: Changing this back reminded me why I changed it in the first place, the base items cause an error because they use  "torso_hanging_back" which should only use the BELTED layer.

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json
Comment (dseguin) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json:20]:
```suggestion
    "name": { "str_sp": "scabbard (exosuit)" },
```

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json
Comment (dseguin) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json:37]:
```suggestion
    "name": { "str": "large storage pack (exosuit)" },
```

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json
Comment (dseguin) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json:51]:
```suggestion
    "name": { "str": "medium pouch (exosuit)", "str_pl": "medium pouches (exosuit)" },
```

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json
Comment (dseguin) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json:66]:
```suggestion
    "name": { "str": "small pouch (exosuit)", "str_pl": "small pouches (exosuit)" },
```

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json
Comment (dseguin) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json:80]:
```suggestion
    "name": { "str_sp": "large ammo pouches (exosuit)" },
```

File: data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json
Comment (dseguin) [inline data/mods/Aftershock/items/armor/modular_exos/modular_exoskeleton_storage.json:95]:
```suggestion
    "name": { "str_sp": "small ammo pouches (exosuit)" },
```

**Review by dseguin — COMMENTED** [review]
Here's some changes necessary for this to work with the management UI:
- Add a "name" field to pocket_data: This is the name used in the UI.
- Add "MANAGE_EXOSUIT" to exoskeleton use_action: This allows the UI to be accessed from the `a`ctivate menu.
- Some minor syntax fixes: Some fields are missing a `"`.

**Comment by John-Candlebury** [comment]
Should be working with the new UI and ready now right? Just confirming before I test it one final time and approve it.

File: data/mods/Aftershock/items/armor/modular_exos/exo_recipe.json
Comment (Drew4484) [reply data/mods/Aftershock/items/armor/modular_exos/exo_recipe.json:1]: Removed the recipes entirely for now, I'll rewrite from the ground up in the next day or so.

**Comment by Drew4484** [comment]
> Should be working with the new UI and ready now right? Just confirming before I test it one final time and approve it.

I haven't had a chance to do detailed testing, hopefully today/tomorrow for that.

**Comment by kevingranade** [comment]
I'm deferring to aftershock maintainers on this, GTG? need any changes?

**Comment by Drew4484** [comment]
> I'm deferring to aftershock maintainers on this, GTG? need any changes?

I need to do a bit more testing tonight and then it should be good to go.

**Comment by John-Candlebury** [comment]
I plan to check it this saturday too! 
The changes are rather extensive so I really want to give them a close look.

**Comment by Drew4484** [comment]
Adjusting some pockets to fit the UI better, instead of large pockets to hold multiple modules.

Helmet doesn't fit in the current UI, we could integrate the helmet into the suit, move the modules to the torso and make it just armor, or rework the UI to incorporate the helmet pockets.

PSU plating can't be loaded via the UI which might be a simple fix. If not, we could always integrate armor into the PSUs to simplify that.

@John-Candlebury other than the notes above, the suits seem to be fine in combat and longer operational times. Let me know if you encounter anything in your tests to fix.

**Comment by Drew4484** [comment]
Realized I made local changes but didn't push them, I'll reconcile with your edits later tonight.

**Comment by John-Candlebury** [comment]
Oh got it.
I still planned to fix a few other issues. Do you want me to do that or should I rather list the things that need fixing?

**Comment by Drew4484** [comment]
> Oh got it.
> I still planned to fix a few other issues. Do you want me to do that or should I rather list the things that need fixing?

Go ahead and push what you find, I'll sort through it tonight.

**Comment by John-Candlebury** [comment]
Feel free to update, Im going to look into fixing the issue linked above,

**Comment by Drew4484** [comment]
#54764 fixes the encumbrance issue, now just need to pass tests and this should be done.

**Comment by John-Candlebury** [comment]
Alright some notes from my  final testing:

-Neither the enhanced optics or the hear protection modules work. Fixing them is actually slightly involved, I suggest you remove them from this PR and submit them in a different one.
-The exosuit repair kit complains about misconfigured battery & magazine data when spawned.
-Theres a conflict in the flags file.

Once those issues are solved this is good for merging,  Thanks for all the work you put into it.

**Comment by John-Candlebury** [comment]
You added a lot of extraneous changes with your last commit. You should undo it using `git reset --hard HEAD~1` from the command line. And then run `git push -f` to remove it from this pr.

Ping me in the discord if you want help with the rebase and merge conflict afterwards.

**Review by John-Candlebury — APPROVED** [review]

**Pull request merged**
